torch/distributed/device_mesh.py

```diff
@@ -344,12 +355,7 @@ def __hash__(self):
         def __eq__(self, other: object) -> bool:
             if not isinstance(other, DeviceMesh):
                 return False
-            if id(self.mesh) == id(other.mesh):
-                return True
-            return (
-                self.mesh.shape == other.mesh.shape
-                and self._flatten_mesh_list == other._flatten_mesh_list
-            )
+            return self._hash == other._hash
```

Comment:
I think we still want to keep the `id(self.mesh) == id(other.mesh)` branch, hashing is equal does not mean it's absolutely equal I think

Reply:
@wanchaol confused - does this need to check identity (in which case you could do `mesh_a is mesh_b`?), or just logical equality?

The hash IIUC already compares all the fields in the latest change.

Reply:
We could possible just do `mesh_a is mesh_b`, this branch is more like a short cut return instead of comparing all fields for perf reason.

hash and eq must agree, so even if hashing hash all the fields, it does not mean that these two object are equal, because two different object have the possibility to hash to the same bucket

Reply:
Yeah don't get me wrong - not saying `hash` is enough as you said collision could happen.

I think my comment was more from an angle of value vs ref equality. Generally I think when ppl do `==` in python they expect the former? Since otherwise they would do `is` directly. So in that sense I think I was more expecting `self.foo == other.foo and self.bar == other.bar` instead of `id()` in `__eq__`.

Reply:
Yeah I think the id check is more like a perf optimization, it have a baked in assumption that each DeviceMesh is immutable after construction (which is true that we don't want DeviceMesh to be mutable). So if we just check `id() == id()` it would be sufficient to know they are the same in memory object thus they should be equal, `a is b` is also aliasing to `id()` check I think. So I think we should keep both logic:

1. check if they are the same object and early return
2. if they are not same object, we do `==` by comparing all fields